lib/client/api.go

```diff
@@ -674,6 +674,20 @@ func WithBeforeLoginHook(fn func() error) RetryWithReloginOption {
 
 // IsErrorResolvableWithRelogin returns true if relogin is attempted on `err`.
 func IsErrorResolvableWithRelogin(err error) bool {
+	// Private key policy errors indicate that the user must login with an
+	// unexpected private key policy requirement satisfied. This can occur
+	// in the following cases:
+	// - User is logging in for the first time, and their strictest private
+	//   key policy requirement is specified in a role.
+	// - User is assuming a role with a stricter private key policy requirement
+	//   than the user's given roles.
+	// - The private key policy in the user's roles or the cluster auth
+	//   preference have been upgraded since the user last logged in, making
+	//   their current login session invalid.
+	if keys.IsPrivateKeyPolicyError(err) {
```

Comment:
Add a test to avoid future regressions?

Reply:
I'm actually going to make a follow up PR to add more hardware key support tests in general using a mock implementation of YubikeyPrivateKey with mocked attestation on the server side. Initially, we didn't have any e2e tests for hardware key support since we couldn't use an actual yubikey for CI tests and I lacked the imagination to mock it all out.

Reply:
Shorter answer, adding a good regression test is a bit out of scope of this PR.

Reply:
Can't we create a PrivateKeyPolicyError and call IsErrorResolvableWithRelogin with it? I'm looking for anything too evolved here.

Reply:
(Glad to know about your Yubikey mock though!)

Reply:
Ah yeah I'll just add a unit test.